README.md
# 🖌️ Brush

> Scala Terminal String Styling

## Features
🪶 Dependency free<br/>
🤹‍♂️ Apply multiple text styles<br/>
🎨 Supports 256-color and Truecolor<br/>
🏳️‍🌈 Gradient text styling<br/>
🕵️‍♂️ Automatically detects color support and style accordingly<br/>

![brush](assets/brush.png)

[![Build Status](https://github.com/ranyitz/brush/actions/workflows/ci.yml/badge.svg)](https://github.com/ranyitz/brush/actions/workflows/ci.yml)
[![Maven Central](https://img.shields.io/maven-central/v/io.github.ranyitz/brush_2.13.svg?label=Maven%20Central)](https://central.sonatype.com/artifact/io.github.ranyitz/brush_2.13)
[![Scala Version](https://img.shields.io/badge/scala-2.12%20%7C%202.13%20%7C%203.0-blue.svg)](https://www.scala-lang.org/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Installation
To include Brush in your Scala project, add the following dependency:

```scala
libraryDependencies += "io.github.ranyitz" %% "brush" % "0.3.0"
```

## Usage
To start styling your terminal strings, simply import `brush._` in your Scala file:

```scala
import brush._
 
"bold".bold
"green".green
"inverted".blue.inverse
"strikethrough".bgYellow.italic.strikethrough
"use any css color".color("darkorchid")
"use a specific rgb color"
    .bgRgb(0, 66, 77)
    .rgb(255, 190, 255)
    .bold
    .underline
"use a specific hex color".hex("#D2691E").bgHex("#E6E6FA"))
"gradient background with rgb colors"
    .gradient(
    (255, 102, 102), // Red
    (255, 204, 102), // Orange
    (255, 255, 102), // Yellow
    (204, 255, 102) // Green
    )
"gradient background with named css colors"
    .bgGradient(
    "lavender",
    "peachpuff",
    "mintcream",
    "aliceblue",
    "palegoldenrod",
    "lightgray",
    "paleturquoise"
    )
    .black
"gradient with hex triplet colors"
    .gradient(
    "#008080", // Teal
    "#FFFF66", // Yellow
    "#FFCC66" // Orange
    ).black
```

Another option if you prefer not to use the implicit API:

```scala
import brush.Brush

val redText = Brush.decorate("red text").red
```
> After creating a decorated string, you can apply any of the colors and modifiers.

### Text Modifiers
Brush provides various text modifiers for customization:

* `bold` - Makes the text bold.
* `dim` - Applies lower opacity to the text.
* `italic` - Renders the text in italics.
* `underline` - Adds a horizontal line below the text.
* `inverse` - Inverts background and foreground colors.
* `strikethrough` - Strikes a horizontal line through the text's center.
* `overline` - Adds a horizontal line above the text.
* `reset` - Resets the current style.
* `hidden` - Prints the text invisibly.

### Text Colors

`black` | `red` | `green` | `yellow` | `blue` | `magenta` | `cyan` | `white` | `blackBright` | `gray` | `redBright` | `greenBright` | `yellowBright` | `blueBright` | `magentaBright` | `cyanBright` | `whiteBright`

### Background Colors

`bgBlack` | `bgRed` | `bgGreen` | `bgYellow` | `bgBlue` | `bgMagenta` | `bgCyan` | `bgWhite` | `bgBlackBright` | `bgGray` | `bgRedBright` | `bgGreenBright` | `bgYellowBright` | `bgBlueBright` | `bgMagentaBright` | `bgCyanBright` | `bgWhiteBright`

### 256 and Truecolor Support

* `rgb` | `truecolor` - choose a color by providing red, green, and blue values between 0 and 255.
* `bgRgb` | `bgTruecolor` - same as rgb, but for background colors.

```scala
println("Yellow Green".rgb(154, 205, 50))
println("Steel Blue".bgRgb(70, 130, 180))
```

### CSS Named Colors

* `css` | `color` - choose a color by providing a CSS color name.
* `bgCss` | `bgColor` - same as css, but for background colors.

```scala
println("(154, 205, 50)".css("yellowgreen"))
println("(70, 130, 180)".bgCss("steelblue"))
```

For a complete list of available colors, refer to the [full named colors list](https://developer.mozilla.org/en-US/docs/Web/CSS/named-color)

### Hex Triplet Colors

* `hex` - choose a color by providing a [hex triplet](https://en.wikipedia.org/wiki/Web_colors#Hex_triplet).
* `bgHex` - same as hex, but for background colors.

```scala
println("Yellow Green".hex("#9acd32"))
println("Steel Blue".bgHex("#4682b4"))
```

### Gradients
> Note: Gradients are only supported in terminals that support 256-colors or Truecolor.

* `gradient` - apply a gradient to the text by providing a list of colors.
* `bgGradient` - same as gradient, but for background colors.

A color can be provided in any of the following formats:
* CSS color name
* Hex triplet
* RGB tuple

```scala
println("Gradient".gradient("red", "orange", "yellow"))
println("Gradient".bgGradient("#ff0000", "#ffa500", "#ffff00"))
println("Gradient".gradient((255, 0, 0), (255, 165, 0), (255, 255, 0)))
```

### Environment Variables
* `NO_COLOR` - If set to any value, disables all colors.
* `FORCE_COLOR` - If set to any value between `0`-`3`, forces the following color level:
    - `0` - All colors disabled.
    - `1` - Basic 16 colors support.
    - `2` - ANSI 256 colors support.
    - `3` - Truecolor 16 million colors support.

### Inspiration
* [chalk](https://github.com/chalk/chalk) (js)
* [colored](https://github.com/colored-rs/colored) (rust)
* [colored](https://github.com/defunkt/colored) (ruby)
* [rainbow](https://github.com/ktoso/scala-rainbow) (scala)


CHANGELOG.md
# Changelog

## 0.3.0 (Sep 30, 2023)
Expose minimal API
Add explicit API usage (Brush.decorate)

## 0.2.1 (Sep 19, 2023)
Add support for hex colors without the `#` prefix
Add full test coverage

## 0.2.0 (Sep 19, 2023)
Add support for hex colors

## 0.1.0 (Sep 10, 2023)
First version

src/main/scala/brush/ColorMap.scala
package brush

private object ColorMap {
  val colorMap = Map(
    "aliceblue" -> (240, 248, 255),
    "antiquewhite" -> (250, 235, 215),
    "aqua" -> (0, 255, 255),
    "aquamarine" -> (127, 255, 212),
    "azure" -> (240, 255, 255),
    "beige" -> (245, 245, 220),
    "bisque" -> (255, 228, 196),
    "black" -> (0, 0, 0),
    "blanchedalmond" -> (255, 235, 205),
    "blue" -> (0, 0, 255),
    "blueviolet" -> (138, 43, 226),
    "brown" -> (165, 42, 42),
    "burlywood" -> (222, 184, 135),
    "cadetblue" -> (95, 158, 160),
    "chartreuse" -> (127, 255, 0),
    "chocolate" -> (210, 105, 30),
    "coral" -> (255, 127, 80),
    "cornflowerblue" -> (100, 149, 237),
    "cornsilk" -> (255, 248, 220),
    "crimson" -> (220, 20, 60),
    "cyan" -> (0, 255, 255),
    "darkblue" -> (0, 0, 139),
    "darkcyan" -> (0, 139, 139),
    "darkgoldenrod" -> (184, 134, 11),
    "darkgray" -> (169, 169, 169),
    "darkgreen" -> (0, 100, 0),
    "darkkhaki" -> (189, 183, 107),
    "darkmagenta" -> (139, 0, 139),
    "darkolivegreen" -> (85, 107, 47),
    "darkorange" -> (255, 140, 0),
    "darkorchid" -> (153, 50, 204),
    "darkred" -> (139, 0, 0),
    "darksalmon" -> (233, 150, 122),
    "darkseagreen" -> (143, 188, 143),
    "darkslateblue" -> (72, 61, 139),
    "darkslategray" -> (47, 79, 79),
    "darkturquoise" -> (0, 206, 209),
    "darkviolet" -> (148, 0, 211),
    "deeppink" -> (255, 20, 147),
    "deepskyblue" -> (0, 191, 255),
    "dimgray" -> (105, 105, 105),
    "dodgerblue" -> (30, 144, 255),
    "firebrick" -> (178, 34, 34),
    "floralwhite" -> (255, 250, 240),
    "forestgreen" -> (34, 139, 34),
    "fuchsia" -> (255, 0, 255),
    "gainsboro" -> (220, 220, 220),
    "ghostwhite" -> (248, 248, 255),
    "gold" -> (255, 215, 0),
    "goldenrod" -> (218, 165, 32),
    "gray" -> (128, 128, 128),
    "green" -> (0, 128, 0),
    "greenyellow" -> (173, 255, 47),
    "honeydew" -> (240, 255, 240),
    "hotpink" -> (255, 105, 180),
    "indianred" -> (205, 92, 92),
    "indigo" -> (75, 0, 130),
    "ivory" -> (255, 255, 240),
    "khaki" -> (240, 230, 140),
    "lavender" -> (230, 230, 250),
    "lavenderblush" -> (255, 240, 245),
    "lawngreen" -> (124, 252, 0),
    "lemonchiffon" -> (255, 250, 205),
    "lightblue" -> (173, 216, 230),
    "lightcoral" -> (240, 128, 128),
    "lightcyan" -> (224, 255, 255),
    "lightgoldenrodyellow" -> (250, 250, 210),
    "lightgray" -> (211, 211, 211),
    "lightgreen" -> (144, 238, 144),
    "lightpink" -> (255, 182, 193),
    "lightsalmon" -> (255, 160, 122),
    "lightseagreen" -> (32, 178, 170),
    "lightskyblue" -> (135, 206, 250),
    "lightslategray" -> (119, 136, 153),
    "lightsteelblue" -> (176, 196, 222),
    "lightyellow" -> (255, 255, 224),
    "lime" -> (0, 255, 0),
    "limegreen" -> (50, 205, 50),
    "linen" -> (250, 240, 230),
    "magenta" -> (255, 0, 255),
    "maroon" -> (128, 0, 0),
    "mediumaquamarine" -> (102, 205, 170),
    "mediumblue" -> (0, 0, 205),
    "mediumorchid" -> (186, 85, 211),
    "mediumpurple" -> (147, 112, 219),
    "mediumseagreen" -> (60, 179, 113),
    "mediumslateblue" -> (123, 104, 238),
    "mediumspringgreen" -> (0, 250, 154),
    "mediumturquoise" -> (72, 209, 204),
    "mediumvioletred" -> (199, 21, 133),
    "midnightblue" -> (25, 25, 112),
    "mintcream" -> (245, 255, 250),
    "mistyrose" -> (255, 228, 225),
    "moccasin" -> (255, 228, 181),
    "navajowhite" -> (255, 222, 173),
    "navy" -> (0, 0, 128),
    "oldlace" -> (253, 245, 230),
    "olive" -> (128, 128, 0),
    "olivedrab" -> (107, 142, 35),
    "orange" -> (255, 165, 0),
    "orangered" -> (255, 69, 0),
    "orchid" -> (218, 112, 214),
    "palegoldenrod" -> (238, 232, 170),
    "palegreen" -> (152, 251, 152),
    "paleturquoise" -> (175, 238, 238),
    "palevioletred" -> (219, 112, 147),
    "papayawhip" -> (255, 239, 213),
    "peachpuff" -> (255, 218, 185),
    "peru" -> (205, 133, 63),
    "pink" -> (255, 192, 203),
    "plum" -> (221, 160, 221),
    "powderblue" -> (176, 224, 230),
    "purple" -> (128, 0, 128),
    "rebeccapurple" -> (102, 51, 153),
    "red" -> (255, 0, 0),
    "rosybrown" -> (188, 143, 143),
    "royalblue" -> (65, 105, 225),
    "saddlebrown" -> (139, 69, 19),
    "salmon" -> (250, 128, 114),
    "sandybrown" -> (244, 164, 96),
    "seagreen" -> (46, 139, 87),
    "seashell" -> (255, 245, 238),
    "sienna" -> (160, 82, 45),
    "silver" -> (192, 192, 192),
    "skyblue" -> (135, 206, 235),
    "slateblue" -> (106, 90, 205),
    "slategray" -> (112, 128, 144),
    "snow" -> (255, 250, 250),
    "springgreen" -> (0, 255, 127),
    "steelblue" -> (70, 130, 180),
    "tan" -> (210, 180, 140),
    "teal" -> (0, 128, 128),
    "thistle" -> (216, 191, 216),
    "tomato" -> (255, 99, 71),
    "turquoise" -> (64, 224, 208),
    "violet" -> (238, 130, 238),
    "wheat" -> (245, 222, 179),
    "white" -> (255, 255, 255),
    "whitesmoke" -> (245, 245, 245),
    "yellow" -> (255, 255, 0),
    "yellowgreen" -> (154, 205, 50)
  )

  def getRgb(color: String): (Int, Int, Int) = {
    this.colorMap.get(color.toLowerCase) match {
      case Some(rgb) => rgb
      case None => throw new Exception(s"Color ${color.magenta} not found")
    }
  }
}
